### DISCUSSION TITLE
유저 부분만 3계층 아키텍처를 적용하고 싶은데 혹시 괜찮을까요?

### DISCUSSION THREAD

**Discussion by ghost** [tastekim/WeAllLie-BE#147] [Q&A] [answered]
유저 부분은 소켓 관련이 아니다보니,
2계층 만으로는 코드 정리가 조금 어려운 것 같습니다. 
최대한 노력해보려고는 했는데 그래도 provider 부분과 repo의 경계가 애매해지는 부분이 생기네요.
유저 폴더 부분만 3계층 적용하고 싶은데 혹시 괜찮을까요?
@kimhsno1 @tastekim  @yunjin5450

**Comment by tastekim** [comment, marked as answer]
### 현재 디렉토리 구조
```
├── src
│   ├── app.js
│   ├── chat
│   │   └── chat-socket.js
│   ├── game
│   │   ├── game-provider.js
│   │   ├── game-repo.js
│   │   └── game-socket.js
│   ├── middlewares
│   │   ├── auth-middleware.js
│   │   ├── error-handler.js
│   │   ├── login-middleware.js
│   │   ├── passport
│   │   └── socket-auth-middleware.js
│   ├── redis.js
│   ├── rooms
│   │   └── room-socket.js
│   ├── rtcChat
│   │   └── rtcChat.js
│   ├── schemas
│   │   ├── game.js
│   │   ├── index.js
│   │   ├── room.js
│   │   └── user.js
│   ├── socket.js
│   └── users
│       ├── jwt.js
│       ├── user-provider.js
│       ├── user-repo.js
│       └── user-route.js
├── test
└── 제목 없는 다이어그램.drawio
```
저희 user 부분이 이미 3계층 레이어드가 적용된 것 아닌가요?  

`Client` -> `Logic` -> `Data`  

이렇게 3가지 패턴으로 나누자면 `Client Tier`는 `user-routes`가, `Logic Tier` 부분은 `user-provider.js` 가, `Data` 부분은 `user-repo.js`가 맡고 있는 것이 아닌가란 생각이 들어서요 ! 뭔가 방금 든 생각이 혹시 클라이언트의 응답/요청 부분이 현재 `user-provider.js` 에 포함되어 있으니까 `user-route.js`를 단순히 라우팅만 받는 부분으로 하고 `user-controller` 를 만들어서 앞 전에 진행했던 프로젝트들 처럼 분리하고 싶으신 것 같은데 맞을까요?

**Reply by ghost** [reply]
네 맞아요! route 에 controller 역할을 추가하기엔 어떤 경로가 있는지 알아보기 어려워서 route는 따로 존재하는 게 좋다고 생각하고, 나머지를 3단계로 구분하고 싶어서요.

controller도 굳이 DB를 거치지 않아도 되는 유저에게 바로 받는 데이터에 대한 검증 로직이 들어갈 수도 있어서 어느 정도 길어질 수도 있고
다른 기수들 깃헙 볼때도 route.js 를 통해서 패스를 한 눈에 파악할 수 있어서 좋았던 것 같아요.

**Reply by tastekim** [reply]
좋은 것 같아요 ! route 까지 구분하면 확실히 더 분명하게 나눌 수 있을 것 같네요 ))
